File: array/prod_of_array_exp_self.cc
#include <vector>
#include <stdio.h>

std::vector<int> calc(std::vector<int> nums) {
    // time:O(N) space:O(1) is accepted
    int N = nums.size();
    std::vector<int> out(N, 1);
    int left = 1;
    int right = 1;
    for (int i = 0; i < N; ++i) {
        out[i] *= left;
        out[N-i-1] *= right;
        left *= nums[i];
        right *= nums[N-i-1];
    }
    return out;
}
int main() {
    std::vector<int> nums{1,2,3,4,5,6};
    auto out = calc(nums);
    for (int i = 0 ; i < out.size(); ++i) {
        printf("%d ", out[i]);
    }
}

File: array/prod_of_array_exp_self.go
package main

import "fmt"

func calc(nums []int) []int {
	out := make([]int, len(nums))
	for i, _ := range out {
		out[i] = 1
	}

	right := 1
	left := 1
	for i := 0; i < len(nums); i++ {
		out[i] *= left
		out[len(nums)-1-i] *= right
		left *= nums[i]
		right *= nums[len(nums)-1-i]
	}
	return out
}

func main() {
	fmt.Println(calc([]int{1, 2, 3, 4, 5, 6}))
}
